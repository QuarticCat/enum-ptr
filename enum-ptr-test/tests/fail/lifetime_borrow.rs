#![allow(dead_code)]

use enum_ptr::{Compact, EnumPtr};

#[derive(EnumPtr)]
#[enum_ptr(borrow, borrow_mut)]
#[repr(C, usize)]
enum Foo<'a, 'b> {
    A(Option<&'a mut i32>),
    B(Option<&'b mut i32>),
}

fn test_get_ref() {
    let foo: Compact<_> = Foo::A(None).into();
    let foo_ref = foo.borrow();
    drop(foo);
    drop(foo_ref);
}

fn test_get_mut() {
    let mut foo: Compact<_> = Foo::A(None).into();
    let foo_mut = foo.borrow_mut();
    drop(foo);
    drop(foo_mut);
}

fn main() {}
