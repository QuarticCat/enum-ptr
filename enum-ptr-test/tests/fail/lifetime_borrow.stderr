error[E0505]: cannot move out of `foo` because it is borrowed
  --> tests/fail/lifetime_borrow.rs:16:10
   |
14 |     let foo: Compact<_> = Foo::A(None).into();
   |         --- binding `foo` declared here
15 |     let foo_ref = foo.borrow();
   |                   --- borrow of `foo` occurs here
16 |     drop(foo);
   |          ^^^ move out of `foo` occurs here
17 |     drop(foo_ref);
   |          ------- borrow later used here

error[E0505]: cannot move out of `foo` because it is borrowed
  --> tests/fail/lifetime_borrow.rs:23:10
   |
21 |     let mut foo: Compact<_> = Foo::A(None).into();
   |         ------- binding `foo` declared here
22 |     let foo_mut = foo.borrow_mut();
   |                   --- borrow of `foo` occurs here
23 |     drop(foo);
   |          ^^^ move out of `foo` occurs here
24 |     drop(foo_mut);
   |          ------- borrow later used here
